# Part 1. Equation of a Slime

How many late days are you using for this assignment?
Answer: 0 days

In [89]:
# Imports section
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression
from sklearn.preprocessing import PolynomialFeatures
from sklearn.model_selection import cross_val_score
from sklearn.pipeline import make_pipeline
from sklearn.preprocessing import StandardScaler
from sklearn.metrics import confusion_matrix, classification_report, accuracy_score
from sklearn.svm import SVC
from sklearn.linear_model import LogisticRegression
from sklearn.neighbors import KNeighborsClassifier
from sklearn.neural_network import MLPClassifier

## 1. Loading the dataset

In [90]:
# Using pandas load the dataset
science_data = pd.read_csv('science_data_large.csv')
# Output the first 15 rows of the data
print(science_data.head(15))
# Display a summary of the table information (number of datapoints, etc.)
print(science_data.info())


    Temperature °C  Mols KCL     Size nm^3
0              469       647  6.244743e+05
1              403       694  5.779610e+05
2              302       975  6.196847e+05
3              779       916  1.460449e+06
4              901        18  4.325726e+04
5              545       637  7.124634e+05
6              660       519  7.006960e+05
7              143       869  2.718260e+05
8               89       461  8.919803e+04
9              294       776  4.770210e+05
10             991       117  2.441771e+05
11             307       781  5.006455e+05
12             206        70  3.145200e+04
13             437       599  5.390215e+05
14             566        75  9.185271e+04
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 3 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   Temperature °C  1000 non-null   int64  
 1   Mols KCL        1000 non-null   int64  
 2   Size nm^3       1000 n

## 2. Splitting the dataset

In [91]:
# Take the pandas dataset and split it into our features (X) and label (y)
X = science_data[['Temperature °C', 'Mols KCL']]
y = science_data['Size nm^3']
# Use sklearn to split the features and labels into a training/test set. (90% train, 10% test)
# For grading consistency use random_state=42 
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.1, random_state=42)

## 3. Perform a Linear Regression

In [92]:
# Use sklearn to train a model on the training set
model = LinearRegression()
model.fit(X_train, y_train)
# Create a sample datapoint and predict the output of that sample with the trained model
sample_datapoint = pd.DataFrame([[500, 700]], columns=['Temperature °C', 'Mols KCL'])
predicted_size = model.predict(sample_datapoint)
# Report the score for that model using the default score function property of the SKLearn model, in your own words (markdown, not code) explain what the score means
model_score = model.score(X_test, y_test)
print(f"R² Score of the model: {model_score:.4f}")
# Extract the coefficients and intercept from the model and write an equation for your h(x) using LaTeX
coefficients = model.coef_
intercept = model.intercept_

print(f"Coefficients (Temperature, Mols KCL): {coefficients}")
print(f"Intercept: {intercept}")

R² Score of the model: 0.8552
Coefficients (Temperature, Mols KCL): [ 866.14641337 1032.69506649]
Intercept: -409391.47958340764


Write the linear equation of a slime: (example equation: $E = mc^2$)

Answer: \[
h(x) = 866.1464 \times \text{Temperature} + 1032.6951 \times \text{Mols KCL} - 409391.4796
\]


Report on score and explain meaning:

Answer: The R² Score of the model is 0.8552. The score shows how good the model is at predicting the size of the slime. It shows us the proportion of variance in the slime size, predicted by the Temperature and Mols. Because the score is 85.52%, we know that the model is pretty good at predicting. 

## 4. Use Cross Validation

In [93]:
# Use the cross_val_score function to repeat your experiment across many shuffles of the data
# For grading consistency use n_splits=5 and random_state=42
scores = cross_val_score(model, X, y, cv=5)
# Report on their finding and their significance
print("Cross-Validation Scores:", scores)
print(f"Average R² Score from Cross-Validation: {scores.mean():.4f}")
print(f"Standard Deviation of R² Scores: {scores.std():.4f}")

Cross-Validation Scores: [0.83918826 0.87051239 0.85871066 0.87202623 0.84364641]
Average R² Score from Cross-Validation: 0.8568
Standard Deviation of R² Scores: 0.0135


Write findings here:

Answer: The individual R² scores for the 5 cross-validation scores show how good the model is at predicting the target variable, slime size, on the 5 subsets.
The average R² score of 85.68% shows that the model is pretty good at predicting across the 5 subsets. 
The standard deviation of R² scores, 0.0135 shows that the model is nearly perfectly consistent across the subsets. There is not much deviation between the models predictions. 

## 5. Using Polynomial Regression

In [94]:
# Using the PolynomialFeatures library perform another regression on an augmented dataset of degree 2
poly = PolynomialFeatures(degree=2)
model_poly = make_pipeline(poly, LinearRegression())
# Perform k-fold cross validation (as above)
scores_poly = cross_val_score(model_poly, X, y, cv=5)
model_poly.fit(X, y)

model_poly_trained = model_poly.named_steps['linearregression']
coefficients_poly = model_poly_trained.coef_
intercept_poly = model_poly_trained.intercept_
# Report on the metrics and output the resultant equation as you did in Part 3.
print("Cross-Validation Scores for Polynomial Regression:", scores_poly)
print(f"Average R² Score from Cross-Validation: {scores_poly.mean():.4f}")
print(f"Standard Deviation of R² Scores: {scores_poly.std():.4f}")
print(f"Intercept: {intercept_poly}")
print(f"Coefficients: {coefficients_poly}")

Cross-Validation Scores for Polynomial Regression: [1. 1. 1. 1. 1.]
Average R² Score from Cross-Validation: 1.0000
Standard Deviation of R² Scores: 0.0000
Intercept: 1.657189568504691e-05
Coefficients: [ 0.00000000e+00  1.20000000e+01 -1.23110504e-07 -1.05648823e-11
  2.00000000e+00  2.85714287e-02]


Write the polynomial equation of a slime: (example equation: $E = mc^2$)

Answer: 
\[
h(x) = 12 \times (\text{Temperature}^2) - 0.0000001231 \times (\text{Mols KCL}^2) - 0.00000000001056488 \times (\text{Temperature} \times \text{Mols KCL}) + 2 \times \text{Temperature} + 0.02857143 \times \text{Mols KCL} + 0.0000165719
\]

Report on the score and interpret:

Answer: The average R² score from Cross-Validation is 1.0000 and the standard deviation of the R² scores is 0.0000. This means the model prefectly predicts the size of the slime each tome.

# Part 2. Chronic Kidney Disease Prediction via Classification

Create code and markdown cells as needed to perform classification and report on your results

In [95]:
# Load the dataset. Then train and evaluate the classification models.
ckd_data = pd.read_csv('ckd_feature_subset.csv')
print(ckd_data.head())

X = ckd_data.drop(columns=['Target_ckd'])  # All columns except the target
y = ckd_data['Target_ckd']

def evaluate_model(model, X, y):
    scores = cross_val_score(model, X, y, cv=5)
    return scores.mean(), scores.std()
models = {
    "Logistic Regression": LogisticRegression(random_state=42, max_iter=1000),
    "Support Vector Machine": SVC(random_state=42),
    "k-Nearest Neighbors": KNeighborsClassifier(),
    "Neural Network": MLPClassifier(random_state=42, max_iter=1000)
}
results = {}

for model_name, model in models.items():
    mean_score, std_score = evaluate_model(model, X, y)
    results[model_name] = {'Mean Score': mean_score, 'Standard Deviation': std_score}

results_df = pd.DataFrame(results).T
print(results_df)


        age        bp      wbcc  appet_poor  appet_good      rbcc  Target_ckd
0  0.688312  0.333333  0.000000           1           0  0.000000           1
1  0.545455  0.333333  0.128319           1           0  0.305085           1
2  0.714286  0.500000  0.238938           1           0  0.186441           1
3  0.688312  0.333333  0.283186           0           1  0.338983           1
4  0.441558  0.333333  0.221239           1           0  0.220339           1
                        Mean Score  Standard Deviation
Logistic Regression       0.856129            0.054236
Support Vector Machine    0.928172            0.059993
k-Nearest Neighbors       0.928172            0.031893
Neural Network            0.941290            0.024040


## Results and Conclusion for Classification Experiments

| Model                | Mean Score (Accuracy) | Standard Deviation |
|----------------------|-----------------------|---------------------|
| Logistic Regression  | 0.8561                | 0.0542              |
| Support Vector Machine (SVM) | 0.9282         | 0.0599              |
| k-Nearest Neighbors (k-NN)   | 0.9282         | 0.0319              |
| Neural Network       | 0.9413                | 0.0240              |

This shows that Neural Network is the best model because it has the highest mean score, 94.13% and lowest standard deviation, 0.0240. Logistic Regression had the lowest accuracy score. 

Experiments with Neural Network

In [98]:
# Experiments with Neural Network.
def evaluate_nn(hidden_layer_sizes, activation, learning_rate_init, max_iter=1000):
    model = MLPClassifier(
        hidden_layer_sizes=hidden_layer_sizes,
        activation=activation,
        learning_rate_init=learning_rate_init,
        max_iter=max_iter,
        random_state=42
    )
    
    mean_score, std_score = evaluate_model(model, X, y)
    return mean_score, std_score


nn_configurations = [
    {"hidden_layer_sizes": (50,), "activation": "relu", "learning_rate_init": 0.001},
    {"hidden_layer_sizes": (100, 50), "activation": "tanh", "learning_rate_init": 0.001},
    {"hidden_layer_sizes": (100, 100), "activation": "relu", "learning_rate_init": 0.01},
    {"hidden_layer_sizes": (150,), "activation": "tanh", "learning_rate_init": 0.01},
    {"hidden_layer_sizes": (50, 50, 50), "activation": "relu", "learning_rate_init": 0.001},
]


nn_results = {}

for i, config in enumerate(nn_configurations, 1):
    mean_score, std_score = evaluate_nn(
        config["hidden_layer_sizes"], 
        config["activation"], 
        config["learning_rate_init"]
    )
    nn_results[f"NN Config {i}"] = {
        "Hidden Layers": config["hidden_layer_sizes"],
        "Activation": config["activation"],
        "Learning Rate": config["learning_rate_init"],
        "Mean Score": mean_score,
        "Standard Deviation": std_score
    }


nn_results_df = pd.DataFrame(nn_results).T

# Displaying the results clearly
from IPython.display import display  
print("Summary of Neural Network Configurations:\n")
display(nn_results_df)

Summary of Neural Network Configurations:



,Hidden Layers,Activation,Learning Rate,Mean Score,Standard Deviation
NN Config 1,"(50,)",relu,0.001,0.94129,0.02404
NN Config 2,"(100, 50)",tanh,0.001,0.960645,0.024554
NN Config 3,"(100, 100)",relu,0.01,0.98043,0.015984
NN Config 4,"(150,)",tanh,0.01,0.960645,0.024554
NN Config 5,"(50, 50, 50)",relu,0.001,0.967527,0.03534


## Results and Conclusion for Neural Network Experiments

The best model was NN Config 3. It has the highest accuracy score, 98.04%, and lowest standard deviation, 0.015. The most important parameter that made the biggest difference in configuration was the number of hidden layers and their size. Here, it had 2 hidden layers of size (100, 100).